chainstate/src/detail/query.rs

```diff
@@ -177,52 +177,79 @@ impl<'a, S: BlockchainStorageRead, V: TransactionVerificationStrategy> Chainstat
         self.chainstate_ref.get_block_height_in_main_chain(id)
     }
 
-    pub fn get_headers(
+    fn get_headers_above(
         &self,
-        locator: Locator,
+        height: BlockHeight,
         header_count_limit: usize,
     ) -> Result<Vec<SignedBlockHeader>, PropertyQueryError> {
         let header_count_limit = BlockDistance::new(
             header_count_limit.try_into().expect("Unreasonable header count limit"),
         );
 
-        // use genesis block if no common ancestor with better block height is found
-        let mut best = BlockHeight::new(0);
-
-        for block_id in locator.iter() {
-            if let Some(block_index) = self.chainstate_ref.get_gen_block_index(block_id)? {
-                if self.chainstate_ref.is_block_in_main_chain(block_id)? {
-                    best = block_index.block_height();
-                    break;
-                }
-            }
-        }
-
         // get headers until either the best block or header limit is reached
         let best_height = self
             .chainstate_ref
             .get_best_block_index()?
-            .expect("best block's height to exist")
+            .ok_or(PropertyQueryError::BestBlockIndexNotFound)?
             .block_height();
 
         let limit = std::cmp::min(
-            (best + header_count_limit).expect("BlockHeight limit reached"),
+            (height + header_count_limit).expect("BlockHeight limit reached"),
             best_height,
         );
 
-        let headers = itertools::iterate(best.next_height(), |iter| iter.next_height())
+        let headers = itertools::iterate(height.next_height(), |iter| iter.next_height())
             .take_while(|height| height <= &limit)
             .map(|height| self.chainstate_ref.get_header_from_height(&height));
         itertools::process_results(headers, |iter| iter.flatten().collect::<Vec<_>>())
     }
 
+    pub fn get_headers(
+        &self,
+        locator: Locator,
+        header_count_limit: usize,
+    ) -> Result<Vec<SignedBlockHeader>, PropertyQueryError> {
+        // use genesis block if no common ancestor with better block height is found
+        let mut best_height = BlockHeight::new(0);
+
+        for block_id in locator.iter() {
+            if let Some(block_index) = self.chainstate_ref.get_gen_block_index(block_id)? {
+                if self.chainstate_ref.is_block_in_main_chain(block_id)? {
+                    best_height = block_index.block_height();
+                    break;
+                }
+            }
+        }
+
+        self.get_headers_above(best_height, header_count_limit)
+    }
+
+    pub fn get_headers_since_fork_point(
```

Comment:
get_headers_not_in_mainchain()?

Generally, we try to avoid using new terms all the time. It can easily create confusion.

Reply:
Also some documentation would be great. I do understand what this function does based on the name, but generally it's not really clear. Either that, or explain even more in the name.